dev-tools/clickana/src/chart.rs

```diff
@@ -0,0 +1,764 @@
+/// of 1 in MiB or GiB in bytes. If the minimum is equal to or less than 1.0,
+/// or the equivalent of 1 once converted from bytes to the expected unit
+/// (e.g. less than or equal to 1048576 if we're using MiB) we'll use 0.0 as
+/// the minimum value as we don't expect any of our charts
+/// to require negative numbers for now.
+fn padded_min_value_raw(unit: Unit, min_value_raw: &f64) -> Result<f64> {
+    let padded_value = match unit {
+        Unit::Count => {
+            if *min_value_raw <= 1.0 {
+                0.0
+            } else {
+                min_value_raw - 1.0
+            }
+        }
+        Unit::Gibibyte | Unit::Mebibyte => {
+            let bytes = unit.as_bytes_f64()?;
+            if *min_value_raw <= bytes {
+                0.0
+            } else {
+                min_value_raw - bytes
+            }
+        }
+    };
+    Ok(padded_value.floor())
+}
+
+/// Returns the difference of the minimum raw value and 1 or the equivalent
+/// of 1 in MiB or GiB in bytes. If the minimum is less than 1, we'll use
+/// 0 as the minimum value as we don't expect any of our charts to require
+/// negative numbers for now.
+fn padded_min_value_as_unit(unit: Unit, min_value_raw: &f64) -> Result<f64> {
+    let padded_value = match unit {
+        Unit::Count => {
+            if *min_value_raw < 1.0 {
+                0.0
+            } else {
+                min_value_raw - 1.0
+            }
+        }
+        Unit::Gibibyte | Unit::Mebibyte => {
+            let value_as_unit = min_value_raw / unit.as_bytes_f64()?;
+            if value_as_unit < 1.0 {
+                0.0
+            } else {
+                value_as_unit - 1.0
+            }
+        }
+    };
+    Ok(padded_value.floor())
+}
+
+#[derive(Debug, PartialEq)]
+struct YAxisValues {
+    lower_label: String,
+    mid_label: String,
+    upper_label: String,
+    lower_bound: f64,
+    upper_bound: f64,
+}
+
+impl YAxisValues {
+    fn new(unit: Unit, raw_data: &Vec<SystemTimeSeries>) -> Result<Self> {
+        // Retrieve values only to create Y axis bounds and labels
+        let values = TimeSeriesValues::new(&raw_data);
+        let max_value = values.max()?;
+        let min_value = values.min()?;
+
+        // In case there is very little variance in the y axis, we will be adding some
+        // padding to the bounds and labels so we don't end up with repeated labels or
+        // straight lines too close to the upper bounds.
+        let upper_bound = padded_max_value_raw(unit, max_value)?;
+        let upper_label_as_unit = padded_max_value_as_unit(unit, max_value)?;
+        let lower_bound = padded_min_value_raw(unit, min_value)?;
+        let lower_label_as_unit = padded_min_value_as_unit(unit, min_value)?;
+        let mid_label_as_unit =
+            values.avg(lower_label_as_unit, upper_label_as_unit);
+
+        // To nicely display the mid value label for the Y axis, we do the following:
```

Comment:
I think you can get rid of this parsing if you just convert all the values to integers at ingestion time.

The only reason I guess you wouldn't want to do this is if there are metrics where there are fractions we actually care about.

Reply:
I get that this looks super weird 😄 , but I added it for the cases when there is very little variance between each point, and the bounds end up being very close to each other. The rounding made the mid point not be mid at all, and the data didn't really match up with the labels anymore